lang/eval.go

```diff
@@ -226,22 +225,63 @@ func (s *Scope) evalContext(refs []*addrs.Reference, selfAddr addrs.Referenceabl
 				continue
 			}
 
-			// Treat "self" as an alias for the configured self address.
-			rawSubj = selfAddr
-			isSelf = true
+			// self can only be used within a resource instance
+			subj := selfAddr.(addrs.ResourceInstance)
 
-			if rawSubj == addrs.Self {
+			if selfAddr == addrs.Self {
 				// Programming error: the self address cannot alias itself.
 				panic("scope SelfAddr attempting to alias itself")
 			}
+
+			val, valDiags := normalizeRefValue(s.Data.GetResource(subj.ContainingResource(), rng))
```

Comment:
Rather than handling this as a totally special case up here, could we instead get the same result by setting `rawSubj` to `subj.ContainingResource` and then just letting this fall out into the switch statement below? If I'm reading this right, it seems like we'd then branch to the `case addrs.Resource` label and get the same result.

(Alternatively, maybe we could just skip populating `managedResources` altogether here; populating `self` should be sufficient I think, because if there were any references to the resource itself as well then we'll catch them on a different iteration of the loop. That would also remove the risk that some future enhancement allows `self` to refer to data resources and we end up incorrectly putting them in the `managedResources` map below, just thinking defensively about what ways outside code might violate these assumptions in future.)

Reply:
I think your latter suggestion is correct, and I just followed existing code+tests which ended up populating the manages resources. 

We need to assign an _instance_ to `self` in the final context, so I think this section is still required to extract the "self" value from the resource as a whole, whereas the switch below is adding resources.